playbooks/roles/mongo_3_2/templates/backup-mongo.sh.j2

```diff
@@ -0,0 +1,67 @@
+#!/bin/bash
+# ref https://tasks.opencraft.com/browse/SE-1669
+# Script to perform a point-in-time dump of the local mongodb database using
+# mongodump.
+# includes locking (prevent this script from running multiple times in
+# parallel), creating a snapshot of the volume used to backup to, and snitch monitoring
+
+# exit by default on failure
+set -e
+
+# vars
+ARCHIVE_NAME="mongo-backup-$(date --iso-8601=minutes --utc).gz"
+VOLUME="{{ MONGO_BACKUP_EBS_VOLUME_DEVICE }}" # eg. /dev/svdk
+VOLUME_ID="{{ MONGO_BACKUP_EBS_VOLUME_ID }}" # eg. vol-123456
+VOLUME_MOUNT_PATH="{{ MONGO_BACKUP_VOLUME_MOUNT_PATH }}" # eg. vol-123456
+SNITCH_URL="{{ MONGO_BACKUP_SNITCH_URL }}"
+MONGO_USER="{{ COMMON_MONGO_READ_ONLY_USER }}"
+MONGO_PASS="{{ COMMON_MONGO_READ_ONLY_PASS }}"
+HOSTNAME="{{ MONGO_BACKUP_HOST }}"
+
+# TODO: is hostname what we want here?
+if [ "$HOSTNAME" = $(hostname) ]; then
+    echo "This is not the backup host. Run on a different instance."
+    exit 1
+fi
+
+# Acquire backup lock using this script itself as the lockfile. If another
+# backup task is already running, then exit immediately.
+exec 200<$0
+flock -n 200 || { echo "Another backup task is already running."; exit 1; }
+
+echo "Starting at $(date)"
+
+# ensure volume is mounted
+mkdir -p "$VOLUME_MOUNT_PATH"
+if ! mountpoint -q "$VOLUME_MOUNT_PATH"; then
+  mount -o discard,defaults,noatime "$MONGO_BACKUP_EBS_VOLUME_DEVICE" "$VOLUME_MOUNT_PATH"
+fi
+
+# clean old backup files to save space and so we start afresh
+rm -rf "$VOLUME_MOUNT_PATH/mongo/"
+
+
+# create the dump
+# XXX: we may want to check how this lays out of disk and how it will play against ebs volume snapshots. The idea was
+# that snapshots would be cheap because you only pay for data blocks that have changed between one snapshot and the
+# next. however, if the mongodump -> gzip -> put in a single archive process ends up not being consistent in layout,
+# this may end up with a lot of the disk content changing between snapshots.
+# TODO: investigate which would be better: a single archive (as below) or a directory with individual files. It's
+# possible that a directory with individual files may be better for performance.
```

Comment:
@pomegranited not sure about these points. :)

Reply:
Yep, good things to look into, and you're welcome also to try it out on stage to see how it actually works out. 

>  The script is pretty much fleshed out now. I was wondering about how we would handle snapshot rotation though. Perhaps it would be better afterall to upload to s3 and put a policy on the bucket (since iirc that is possible)? idk how complex it is to handle multipart s3 uploads from the cli though.

(Bringing this question in here since it's relevant to the above.)

Similarly for S3 -- I recommended EBS volumes because I figured they'd be a) faster and b) cheaper to write to.  But it might be that there's an S3 option (there's a few storage mechanisms) that are also cost-effective.  S3 costs are hard to work out though, since they're dependent on access.  But they can also be self-versioning, which can save some time.

For the EBS volume lifecycle stuff, that's going to have to be in terraform, cf https://www.terraform.io/docs/providers/aws/r/dlm_lifecycle_policy.html

Reply:
> For the EBS volume lifecycle stuff, that's going to have to be in terraform, cf https://www.terraform.io/docs/providers/aws/r/dlm_lifecycle_policy.html

Strike that..  as you've mentioned previously, we need to trigger the snapshot when this backup process completes.  Will look into this.

Reply:
The `awscli` supports creating ebs snapshots, and it is installed by the `aws` role, though I'm not sure where to, or which virtualenv you'll need to activate to use it.  See:

* https://docs.aws.amazon.com/cli/latest/reference/ec2/create-snapshot.html
* [roles/aws/tasks/main.yml](https://github.com/edx/configuration/blob/master/playbooks/roles/aws/tasks/main.yml#L57-L69)
* [awscli requirement](https://github.com/edx/configuration/blob/b2805e094ac3b932f83f8d5b7a84cf3da06256b1/playbooks/roles/aws/templates/requirements.txt.j2#L7)

Reply:
yep :smile:  The only thing left to sort out is how to manage rotating snapshots. Perhaps awscli can also do that.

Reply:
Also note -- I used the [mongo_3_4.yml](https://github.com/edx-olive/configuration/blob/opencraft-release/ginkgo.2-campus/playbooks/mongo_3_4.yml) playbook to deploy mongo to the other stage mongo servers, so you can do the same here.

Reply:
@pomegranited How can we set configuration for the mongo server differently on the new instance?

Reply:
@swalladge  Using the `MONGO_RS_CONFIG` variable which is used by [mongodb_rs_config](https://github.com/edx-olive/configuration/blob/opencraft-release/ginkgo.2-campus/playbooks/library/mongodb_rs_config#L129) to configure the replicaset , e.g.

```yaml
MONGO_RS_CONFIG:
  _id: '{{ MONGO_REPL_SET }}'
  members:
    - host: '...'
    - host: {{MONGO_BACKUP_NODE}}
       hidden: true
```

Reply:
Hmm that script is about discovering information about an existing mongo replica set. That doesn't do any configuration.

Reply:
Nvm, that RS_CONFIG is used by `mongodb_replica_set` to do the actual configuration. found the docs now. :)